engine/sources/blob/blobsource.go

```diff
@@ -133,5 +134,28 @@ func (bs *blobSourceInstance) Snapshot(ctx context.Context, _ session.Group) (ca
 		// TODO: could also include description of original blob source by passing along more metadata
 		cache.WithDescription(fmt.Sprintf("dagger blob source for %s", bs.id.Digest)),
 	}
-	return bs.cache.GetByBlob(ctx, bs.id.Descriptor, nil, opts...)
+	ref, err := bs.cache.GetByBlob(ctx, bs.id.Descriptor, nil, opts...)
+	var needsRemoteProviders cache.NeedsRemoteProviderError
+	if errors.As(err, &needsRemoteProviders) {
```

Comment:
What does this bit do? Can't tell if the PR description references it, but it looks important.

Reply:
It's not actually *strictly* necessary right now, but I made it during earlier attempts at fixing the problem and left it in because it at minimum is harmless and *could* be needed someday.

Basically, it just handles the case where there are lazy blobs we are trying to get and they have been provided by an earlier `CacheMap` call in the solver or in general any caller that included them in the context. You can see the pretty much exact same code in other places in Buildkit like [here](https://github.com/sipsma/buildkit/blob/c3c65787b5e2c2c9fcab1d0b9bd1884a37384c90/worker/base/worker.go#L283).

Right now, there aren't even any lazy blobs being passed through the context from the solver, so it didn't end up fixing the problem. But in theory that *could* happen someday, so it is preferable to just have this code in place in case it's ever needed. For example, I left [this `NOTE:` about a performance optimization we might consider](https://github.com/dagger/dagger/pull/5885/files#diff-4c74453e943c306a535b081c57d1367e0b4b75076940fe141133e94a5e573122R110-R112), which would require this code be in place. There's also a potential need for this if we ever want to use stargz or other remote snapshotters too.